include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,338 @@
+                  {
+                    idxQ = i;
+                  }
+                }
+
+                return CrowdingOperator(idxP,
+                                        idxQ,
+                                        ranks,
+                                        crowdingDistance);
+              }
+    );
+
+    // yeild new population P_{t+1}
+    Info << "NSGA2::Optimize() Get P(t+1)" << std::endl;
+    population.resize(populationSize);
+  }
+
+  std::vector<MatType> bestFront;
+
+  for(size_t f: fronts[0])
+  {
+    bestFront.push_back(population[f]);
+  }
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+  return bestFront;
+}
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType>
+inline void NSGA2::EvaluateObjectives(std::vector<MatType> population,
+                                      MultiobjectiveFunctionType objectives,
+                                      std::vector<arma::vec>& calculatedObjectives)
+{
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    calculatedObjectives[i] = objectives.Evaluate(population[i]);
+  }
+}
+
+template<typename MatType>
+inline void NSGA2::BinaryTournamentSelection(std::vector<MatType>& population)
+{
+  std::vector<MatType> children;
+
+  while (children.size() < population.size())
+  {
+    size_t indexA = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+    size_t indexB = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+
+    if (indexA == indexB)
+    {
+      if (indexB < populationSize - 1)
+        indexB++;
+      else
+        indexB--;
+    }
+
+    MatType parentA = population[indexA];
+    MatType parentB = population[indexB];
+
+    MatType childA = parentA, childB = parentB;
+
+    Info << "NSGA2::BinaryTournamentSelection() Crossover" << std::endl;
+    Crossover(childA, childB, parentA, parentB);
+
+    Info << "NSGA2::BinaryTournamentSelection() Mutate(A)" << std::endl;
+    Mutate(childA);
+    Info << "NSGA2::BinaryTournamentSelection() Mutate(B)" << std::endl;
+    Mutate(childB);
+
+    children.push_back(childA);
+    children.push_back(childB);
+  }
+
+  population.reserve(population.size() + children.size());
+  population.insert(std::end(population), std::begin(children), std::end(children));
+}
+
+template<typename MatType>
+inline void NSGA2::Crossover(MatType& childA,
+                MatType& childB,
+                MatType parentA,
+                MatType parentB)
+{
+  // crossover indices
+  auto idx = arma::randu<MatType>(childA.n_rows, childA.n_cols) < crossoverProb;
+
+  childA = parentA % idx + parentB % (1 - idx);
+  childB = parentA % (1 - idx) + parentA % idx;
+}
+
+template<typename MatType>
+inline void NSGA2::Mutate(MatType& child)
+{
+  child += (arma::randu<MatType>(child.n_rows, child.n_cols) < mutationProb) %
+           (mutationStrength * arma::randn<MatType>(child.n_rows, child.n_cols));
```

Comment:
When lines wrap, the next line should be two tabs from where the previous line began.